File: OTEAServer/Controllers/SubSubAmbitsController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using OTEAServer.Misc;
using OTEAServer.Models;

namespace OTEAServer.Controllers
{
    /// <summary>
    /// Controller class for subsubambits operations
    /// Author: Pablo Ahíta del Barrio
    /// Version: 1
    /// </summary>
    [ApiController]
    [Route("SubSubAmbits")]
    public class SubSubAmbitsController : Controller
    {
        /// <summary>
        /// Database context
        /// </summary>
        private readonly DatabaseContext _context;

        /// <summary>
        /// Class constructor
        /// </summary>
        /// <param name="context">Database context</param>
        public SubSubAmbitsController(DatabaseContext context)
        {
            _context = context;
        }

        /// <summary>
        /// Method that obtains from the database all the subSubAmbits
        /// </summary>
        /// <returns>All subAmbits</returns>
        [HttpGet("all")]
        [Authorize]
        public IActionResult GetAll([FromHeader] string Authorization)
        {
            try
            {
                var subSubAmbits = _context.SubSubAmbits.ToList();
                return Ok(subSubAmbits);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains from the database all the subSubAmbits of an subAmbit
        /// </summary>
        /// <param name="idSubAmbit">Subambit identifier</param>
        /// <param name="idAmbit">Ambit identifier</param>
        /// <returns>All subAmbits</returns>
        [HttpGet("allBySubAmbit")]
        [Authorize]
        public IActionResult GetAllBySubAmbit(int idSubAmbit, int idAmbit, [FromHeader] string Authorization)
        {
            try
            {
                var subSubAmbits = _context.SubSubAmbits.Where(ss => ss.idSubAmbit == idSubAmbit && ss.idAmbit == idAmbit).ToList();
                return Ok(subSubAmbits);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains from the database a subSubAmbit using its identifier
        /// </summary>
        /// <param name="id">subSubAmbit identifier</param>
        /// <returns>The subSubAmbit if exists, null if not.</returns>

        [HttpGet("get")]
        [Authorize]
        public ActionResult<SubSubAmbit> Get([FromQuery] int id, [FromHeader] string Authorization)
        {
            try
            {
                var subSubAmbit = _context.SubSubAmbits.FirstOrDefault(a => a.idSubSubAmbit == id);

                if (subSubAmbit == null)
                    return NotFound();

                return subSubAmbit;
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that appends a new subSubAmbit to the database
        /// </summary>
        /// <param name="subSubAmbit">SubSubAmbit to append</param>
        /// <returns>SubSubAmbit append, null if not</returns>
        [HttpPost]
        [Authorize(Policy = "Administrator")]
        public IActionResult Create([FromBody] SubSubAmbit subSubAmbit, [FromHeader] string Authorization)
        {
            try
            {
                _context.SubSubAmbits.Add(subSubAmbit);
                _context.SaveChanges();
                return CreatedAtAction(nameof(Get), new { idSubSubAmbit = subSubAmbit.idSubSubAmbit }, subSubAmbit);

            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }


        /// <summary>
        /// Method that updates the subSubAmbit
        /// </summary>
        /// <param name="idSubSubAmbit">SubSubAmbit identifier</param>
        /// <param name="subSubAmbit">New subSubAmbit</param>
        /// <returns>subSubAmbit if was updated, null if not</returns>
        [HttpPut]
        [Authorize(Policy = "Administrator")]
        public IActionResult Update([FromQuery] int idSubSubAmbit, [FromBody] SubSubAmbit subSubAmbit, [FromHeader] string Authorization)
        {
            try
            {
                if (idSubSubAmbit != subSubAmbit.idSubSubAmbit)
                    return BadRequest();

                var existingSubSubAmbit = _context.SubSubAmbits.FirstOrDefault(a => a.idSubSubAmbit == idSubSubAmbit);
                if (existingSubSubAmbit is null)
                    return NotFound();


                existingSubSubAmbit.idSubSubAmbit = subSubAmbit.idSubSubAmbit;
                existingSubSubAmbit.idSubAmbit = subSubAmbit.idSubAmbit;
                existingSubSubAmbit.idAmbit = subSubAmbit.idAmbit;
                existingSubSubAmbit.descriptionEnglish = subSubAmbit.descriptionEnglish;
                existingSubSubAmbit.descriptionSpanish = subSubAmbit.descriptionSpanish;
                existingSubSubAmbit.descriptionFrench = subSubAmbit.descriptionFrench;
                existingSubSubAmbit.descriptionBasque = subSubAmbit.descriptionBasque;
                existingSubSubAmbit.descriptionCatalan = subSubAmbit.descriptionCatalan;
                existingSubSubAmbit.descriptionDutch = subSubAmbit.descriptionDutch;
                existingSubSubAmbit.descriptionGalician = subSubAmbit.descriptionGalician;
                existingSubSubAmbit.descriptionGerman = subSubAmbit.descriptionGerman;
                existingSubSubAmbit.descriptionItalian = subSubAmbit.descriptionItalian;
                existingSubSubAmbit.descriptionPortuguese = subSubAmbit.descriptionPortuguese;
                _context.SaveChanges();

                return Ok(existingSubSubAmbit);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that deletes the subSubAmbit using its identifier
        /// </summary>
        /// <param name="id">SubSubAmbit identifier</param>
        /// <returns>Deleted subSubAmbit if deletion was successful, null if not</returns>
        [HttpDelete]
        [Authorize(Policy = "Administrator")]
        public IActionResult Delete([FromQuery] int id, [FromHeader] string Authorization)
        {
            try
            {
                var subSubAmbit = _context.SubSubAmbits.FirstOrDefault(a => a.idSubSubAmbit == id);

                if (subSubAmbit is null)
                    return NotFound();

                _context.SubSubAmbits.Remove(subSubAmbit);
                _context.SaveChanges();

                return NoContent();
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
            
        }
    }
}

File: connection/src/main/java/otea/connection/controller/SubSubAmbitsController.java
package otea.connection.controller;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import cli.indicators.SubSubAmbit;
import otea.connection.ConnectionClient;
import otea.connection.api.SubSubAmbitsApi;
import retrofit2.Call;
import retrofit2.Response;
import session.Session;

/**
 * Controller class for subSubAmbit operations
 *
 * @author Pablo Ahíta del Barrio
 * @version 1
 * */
public class SubSubAmbitsController {

    /**SubSubAmbits api to connect to the server*/
    private static SubSubAmbitsApi api;

    /**SubSubAmbitsController instance for singleton*/
    private static SubSubAmbitsController instance;

    /**Class constructor*/
    private SubSubAmbitsController(){
        api= ConnectionClient.getInstance().getRetrofit().create(SubSubAmbitsApi.class);
    }

    /**
     * Synchronized method that obtains the instance
     *
     * @return Instance of subSubAmbit
     * */
    public static SubSubAmbitsController getInstance(){
        if(instance==null){
            synchronized (AmbitsController.class){
                if(instance==null){
                    instance=new SubSubAmbitsController();
                }
            }
        }
        return instance;
    }

    /**Refresh API*/
    public static void refreshApi(){
        instance=new SubSubAmbitsController();
    }

    /**
     * Method that obtains from the database all the subSubAmbits
     *
     * @return All subSubAmbits
     * */
    public static List<SubSubAmbit> GetAll(){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<SubSubAmbit>> callable = new Callable<List<SubSubAmbit>>() {
            @Override
            public List<SubSubAmbit> call() throws Exception {
                Call<List<SubSubAmbit>> call = api.GetAll(Session.getInstance().getToken());
                Response<List<SubSubAmbit>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<SubSubAmbit>> future = executor.submit(callable);
            List<SubSubAmbit> list = future.get();
            executor.shutdown();
            return list;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that obtains from the database all the subSubAmbits
     *
     * @param idSubAmbit - subAmbit identifier
     * @param idAmbit - Ambit identifier
     * @return All subSubAmbits
     * */
    public static List<SubSubAmbit> GetAllBySubAmbit(int idSubAmbit,int idAmbit){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<SubSubAmbit>> callable = new Callable<List<SubSubAmbit>>() {
            @Override
            public List<SubSubAmbit> call() throws Exception {
                Call<List<SubSubAmbit>> call = api.GetAllBySubAmbit(idSubAmbit, idAmbit, Session.getInstance().getToken());
                Response<List<SubSubAmbit>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<SubSubAmbit>> future = executor.submit(callable);
            List<SubSubAmbit> list = future.get();
            executor.shutdown();
            return list;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that obtains from the database a subSubAmbit using its identifier
     *
     * @param id - subSubAmbit identifier
     * @return The subSubAmbit if exists, null if not.
     * */

    public static SubSubAmbit Get(int id){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<SubSubAmbit> callable = new Callable<SubSubAmbit>() {
            @Override
            public SubSubAmbit call() throws Exception {
                Call<SubSubAmbit> call = api.Get(id,Session.getInstance().getToken());
                Response<SubSubAmbit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<SubSubAmbit> future = executor.submit(callable);
            SubSubAmbit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that appends a new subSubAmbit to the database
     *
     * @param subSubAmbit - SubSubAmbit to append
     * @return SubSubAmbit append, null if not
     * */
    public static SubSubAmbit Create(SubSubAmbit subSubAmbit){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<SubSubAmbit> callable = new Callable<SubSubAmbit>() {
            @Override
            public SubSubAmbit call() throws Exception {
                Call<SubSubAmbit> call = api.Create(subSubAmbit,Session.getInstance().getToken());
                Response<SubSubAmbit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<SubSubAmbit> future = executor.submit(callable);
            SubSubAmbit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that updates the subSubAmbit by the given
     *
     * @param idSubSubAmbit - SubSubAmbit identifier
     * @param subSubAmbit - new subSubAmbit
     * @return subSubAmbit if was updated, null if not
     *
     * */
    public static SubSubAmbit Update(int idSubSubAmbit, SubSubAmbit subSubAmbit){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<SubSubAmbit> callable = new Callable<SubSubAmbit>() {
            @Override
            public SubSubAmbit call() throws Exception {
                Call<SubSubAmbit> call = api.Update(idSubSubAmbit,subSubAmbit,Session.getInstance().getToken());
                Response<SubSubAmbit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<SubSubAmbit> future = executor.submit(callable);
            SubSubAmbit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that deletes the subSubAmbit using its
     *
     * @param id - SubSubAmbit identifier
     * @return Deleted subSubAmbit if deletion was succesful, null if not
     * */
    public static SubSubAmbit Delete(int id){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<SubSubAmbit> callable = new Callable<SubSubAmbit>() {
            @Override
            public SubSubAmbit call() throws Exception {
                Call<SubSubAmbit> call = api.Delete(id,Session.getInstance().getToken());
                Response<SubSubAmbit> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<SubSubAmbit> future = executor.submit(callable);
            SubSubAmbit result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }
}
